Add 5 part of accessing Kafka blog post series (#65)
strimzi/strimzi.github.io: _posts/2019-05-21-accessing-kafka-part-5.md

---
layout: post

> title:  "Accessing Kafka: Part 4 - Load Balancers"

Reviewer: Wrong title :)

date: 2019-05-13
author: jakub_scholz
---

In the fifth and last part of this blog post series we will look at exposing Kafka using Kubernetes Ingress.

> This post will explain how to use Ingress controllers on Kubernetes, how does Ingress compare with OpenShift Routes and how it can be used with Strimzi and Kafka.

Reviewer 2: ```suggestion
This post will explain how to use Ingress controllers on Kubernetes, how Ingress compares with OpenShift Routes and how it can be used with Strimzi and Kafka.
```

> Off-cluster access using Kubernetes Ingress is available only from Strimzi 0.12.0.

Reviewer: I guess we cannot publish the article until this release will be out.

Author: I don't think so - it is just a preview. Also the other posts mentioned some un-released features.

<!--more-->

_This post is part of a bigger series about different ways to access a Kafka cluster powered by Strimzi.
The other parts published so far are:_

* _[Part 1 - Introduction](https://strimzi.io/2019/04/17/accessing-kafka-part-1.html)_
* _[Part 2 - Node Ports](https://strimzi.io/2019/04/23/accessing-kafka-part-2.html)_
* _[Part 3 - OpenShift Routes](https://strimzi.io/2019/04/30/accessing-kafka-part-3.html)_
* _[Part 4 - Load Balancers](https://strimzi.io/2019/05/13/accessing-kafka-part-4.html)_
* _Part 5 - Ingress (this post)_

_This is the last post in the series._

## Kubernetes Ingress

Ingress is a Kubernetes API for managing external access to HTTP/HTTPS services which was added in Kubernetes 1.1.

> Ingress is the Kubernetes counter part to the OpenShift Routes which we discussed in [part 3](https://strimzi.io/2019/04/30/accessing-kafka-part-3.html).

Reviewer 2: ```suggestion
Ingress is the Kubernetes counterpart to OpenShift Routes, which we discussed in [part 3](https://strimzi.io/2019/04/30/accessing-kafka-part-3.html).
```

It acts as a [Layer 7](https://en.wikipedia.org/wiki/OSI_model#Layer_7:_Application_Layer) load balancer for HTTP or HTTPS traffic.
The Ingress resources will define the rules for routing the traffic to different services and pods.

> And an Ingress controller will take care of the actual routing.

Reviewer 2: ```suggestion
An Ingress controller takes care of the actual routing.
```

For more information about Ingress, check the [Kubernetes website](https://kubernetes.io/docs/concepts/services-networking/ingress/).

> Ingress is a bit strange part of the Kubernetes API.

Reviewer 2: ```suggestion
Ingress is a bit of a strange part of the Kubernetes API.
```

> The Ingress API it self is part of every Kubernetes cluster.

Reviewer 2: ```suggestion
The Ingress API itself is part of every Kubernetes cluster,
```

> But the Ingress controller which would do the routing is not part of core Kubernetes.

Reviewer 2: ```suggestion
but the Ingress controller which would do the routing is not part of core Kubernetes.
```

So while you will be able to create the Ingress resources, there might be nothing to actually route the traffic.

So for the Ingress resources to actually do something, you need to make sure an Ingress controller is installed.
There are many different Ingress controllers.

> The Kubernetes project it self has two controllers: 

Reviewer 2: ```suggestion
The Kubernetes project itself has two controllers: 
```

* [NGINX controller](https://github.com/kubernetes/ingress-nginx)
* [GCE controller](https://github.com/kubernetes/ingress-gce) for Google Cloud

But there are many additional controllers created and maintained by different communities and companies.
A list of different controllers can be found on the [Kubernetes website](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/).

Most of the controllers rely on a load balancer or node port service which will get the external traffic to the controller.
Once the traffic reaches the controller, the controller will route it based in the rules specified in the Ingress resource to the different services and pods.

> The controller it self usually also runs as yet another application inside the Kubernetes cluster.

Reviewer 2: ```suggestion
The controller itself usually also runs as yet another application inside the Kubernetes cluster.
```

![Accessing applications using NGINX Ingress Controller]({{ "/assets/2019-05-21-ingress-controller.png" }})

> Some of the controllers are tailored for a specific public clouds - for example the [AWS ALB Ingress Controller](https://github.com/kubernetes-sigs/aws-alb-ingress-controller) is provisioning the AWS Application Load Balancer to do the routing instead of doing it inside some pod in your Kubernetes cluster.

Reviewer 2: ```suggestion
Some of the controllers are tailored for a specific public cloud - for example the [AWS ALB Ingress Controller](https://github.com/kubernetes-sigs/aws-alb-ingress-controller) provisions the AWS Application Load Balancer to do the routing instead of doing it inside some pod in your Kubernetes cluster.
```

Ingress offers a lot of functionality for HTTP applications such as:
* TLS termination
* Redirecting from HTTP to HTTPS
* Routing based on HTTP request path

> Some of the controllers such as the NGINX controller also offer TLS passthrough which we will use in Strimzi.

Reviewer 2: ```suggestion
Some of the controllers such as the NGINX controller also offer TLS passthrough, which is a feature we use in Strimzi.
```

## Using Ingress in Strimzi

Ingress support in Strimzi has been added in Strimzi 0.12.0.

> It is using TLS passthrough and it was tested with the [NGINX Ingress Controller](https://github.com/kubernetes/ingress-nginx).

Reviewer 2: ```suggestion
It uses TLS passthrough and it was tested with the [NGINX Ingress Controller](https://github.com/kubernetes/ingress-nginx).
```

Before using it, please make sure that the TLS passthrough is [enabled in the controller](https://kubernetes.github.io/ingress-nginx/user-guide/tls/#ssl-passthrough).

> Ingress support in Strimzi 0.12.0 can be called _experimental_.

Reviewer 2: ```suggestion
Ingress support in Strimzi 0.12.0 is _experimental_.
```

If you have any feedback to it or if you want to help to make it work with different Ingress controllers, get in touch with us through our [Slack, mailing list or GitHub](https://strimzi.io/contributing/).

Although some Ingress controllers also support working directly with TCP connections, TLS passthrough seems to be more widely supported.
Therefore we decided to prefer TLS passthrough in Strimzi.

> That also means that when using Ingress, TLs encryption will be always enabled.

Reviewer 2: ```suggestion
That also means that when using Ingress, TLS encryption will always be enabled.
```

One of the main differences compared to OpenShift Routes is that for Ingress you have to specify the host address in your Kafka custom resource.
The Router in OpenShift will automatically assign a host address based on the route name and the project.

> But in Ingress it has to be specified in the Ingress resource.

Reviewer 2: ```suggestion
But in Ingress the host address has to be specified in the Ingress resource.
```

> You also have to take care that the host address DNS resolves to the ingress controller.

Reviewer 2: ```suggestion
You also have to take care that DNS resolves the host address to the ingress controller.
```

Strimzi cannot generate it for you, because it does not know which DNS addresses are configured for the Ingress controller.

If you want to try it for example on Minikube or in other environments where you don't have any managed DNS service to add the hosts for the Kafka cluster, you can use one of the wildcard DNS services such as [nip.io](https://nip.io/) or [xip.io](http://xip.io/) and set it to point to the IP address of your Ingress controller.
For example `broker-0.<minikube-ip-address>.nip.io`.

![Kafka clients connecting through Ingress controller]({{ "/assets/2019-05-21-connecting-through-ingress" }})

> The way Strimzi uses Ingress to expose Kafka should be already familiar to you from the previous blog posts.

Reviewer 2: ```suggestion
The way Strimzi uses Ingress to expose Kafka should already be familiar to you from the previous blog posts.
```

> We create one service as a bootstrap service and and additional services for individual access to each of the Kafka brokers in your cluster.

Reviewer 2: ```suggestion
We create one service as a bootstrap service and additional services for individual access to each of the Kafka brokers in your cluster.
```

For each of these services we will also create one Ingress resource with the corresponding TLS passthrough rule.

![Accessing Kafka using Ingress]({{ "/assets/2019-05-21-ingress-access.png" }})

When configuring Strimzi to use Ingress, you have to specify the type of the external listener as `ingress` and specify the ingress hosts used for the different brokers as well as for bootstrap in the `configuration` field:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    listeners:
      # ...
      external:
        type: ingress
        configuration:
          bootstrap:
            host: bootstrap.192.168.64.46.nip.io
          brokers:
          - broker: 0
            host: broker-0.192.168.64.46.nip.io
          - broker: 1
            host: broker-1.192.168.64.46.nip.io
          - broker: 2
            host: broker-2.192.168.64.46.nip.io
    # ...
```

> Using Ingress in your clients is very similar to openShift Routes.

Reviewer: Capital "O" in openShift

Since it always uses TLS Encryption, you have to first download the server certificate (replace `my-cluster` with the name of your cluster):

```
oc extract secret/my-cluster-cluster-ca-cert --keys=ca.crt --to=- > ca.crt
keytool -import -trustcacerts -alias root -file ca.crt -keystore truststore.jks -storepass password -noprompt

> ```

Reviewer: Because Ingress is a thing related to Kubernetes as Routes to OpenShift shouldn't we use the kubectl way to get the certificate? And not oc tool?

Author: Good point. Its shame `kubectl` doesn't have the `extract` command.

Once you have the TLS certificate, you can use the bootstrap host you specified in the Kafka custom resource and connect to the Kafka cluster.

> Since Ingress is using TLS passthrough, you always have to connect on port `443`.

Reviewer 2: ```suggestion
Since Ingress uses TLS passthrough, you always have to connect on port `443`.
```

Using the DNS annotations is simple:

```yaml
# ...
listeners:
  external:
    type: ingress
    configuration:
      bootstrap:
        host: kafka-bootstrap.mydomain.com
      brokers:
      - broker: 0
        host: kafka-broker-0.mydomain.com
      - broker: 1
        host: kafka-broker-1.mydomain.com
      - broker: 2
        host: kafka-broker-2.mydomain.com
    overrides:
      bootstrap:
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-bootstrap.mydomain.com.
          external-dns.alpha.kubernetes.io/ttl: "60"
      brokers:
      - broker: 0
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-0.mydomain.com.
          external-dns.alpha.kubernetes.io/ttl: "60"
      - broker: 1
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-1.mydomain.com.
          external-dns.alpha.kubernetes.io/ttl: "60"
      - broker: 2
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-2.mydomain.com.
          external-dns.alpha.kubernetes.io/ttl: "60"
# ...
```

Again, Strimzi lets you configure the annotations directly.

> That gives you more freedom and hopefully makes this feature usable even when you use some other tool then External DNS.

Reviewer 2: ```suggestion
That gives you more freedom and hopefully makes this feature usable even when you use some other tool than External DNS.
```

It also lets you configure other options than just the DNS names, such as the time-to-live of the DNS records etc.

## Pros and cons

Kubernetes Ingress is not always easy to use because you have to install the Ingress controller and you have to configure the hosts.
It is also available only with TLS encryption because of the TLS passthrough functionality which Strimzi uses.

> But it can offer an interesting option for clusters where node ports are not an option for example for security reasons and using load balancers would be too expensive.

Reviewer 2: ```suggestion
But it can offer an interesting option for clusters where node ports are not an option, for example for security reasons, and where using load balancers would be too expensive.
```

> When using Strimzi Kafka operator with Ingress you always have to be careful about the performance.

Reviewer 2: ```suggestion
When using Strimzi Kafka operator with Ingress you always have to consider performance.
```

> The ingress controller usually runs inside your cluster as yet another deployment and adds additional step through which your data have to flow between our clients and the brokers.

Reviewer 2: ```suggestion
The ingress controller usually runs inside your cluster as yet another deployment and adds an additional step through which your data has to flow between your clients and the brokers.
```

You also need to scale it properly to ensure it will not be a bottleneck for your clients.

So Ingress might not be the best option when most of your applications using Kafka are outside of your Kubernetes cluster and you need to handle 10s or 100s MBs of throughput per second.

> However, especially in situations when most of your applications are inside your cluster and only minority outside and the throughput you need is not so big, Ingress might be a convenient option.

Reviewer 2: ```suggestion
However, especially in situations when most of your applications are inside your cluster and only a minority outside and when the throughput you need is not so high, Ingress might be a convenient option.
```

> The Ingress API and the Ingress controllers can be usually installed on OpenShift clusters as well.

Reviewer 2: ```suggestion
The Ingress API and the Ingress controllers can usually be installed on OpenShift clusters as well.
```

But they do not offer any advantages over the OpenShift Routes.
So on OpenShift, you should prefer the OpenShift Routes instead.

## What's next?

> This was for now the last post in this series about accessing Strimzi Kafka clusters.

Reviewer 2: ```suggestion
This was, for now, the last post in this series about accessing Strimzi Kafka clusters.
```

> In the 5 blog posts, we covered all the supported mechanisms how Strimzi operator allows you to access Kafka from inside and outside of your Kubernetes or OpenShift cluster.

Reviewer 2: ```suggestion
In the 5 blog posts, we covered all the supported mechanisms that the Strimzi operator supports for accessing Kafka from both inside and outside of your Kubernetes or OpenShift cluster.
```

> We will of course keep posting blog posts on another topic and if something about accessing Kafka changes in the future, we will add new blog posts to this series.

Reviewer 2: ```suggestion
We will, of course, keep posting blog posts on other topics and if something about accessing Kafka changes in the future, we will add new blog posts to this series.
```
